snowshu/core/graph.py

```diff
@@ -105,65 +100,142 @@ def _apply_specifications(
                         database=val.database_pattern,
                         schema=val.schema_pattern,
-                        relation=val.relation_pattern,
+                        name=val.relation_pattern,
                         remote_attribute=val.remote_attribute,
-                        local_attribute=val.local_attribute) for val in relation.relationships.__dict__[direction]]
+                        local_attribute=val.local_attribute
+                    ) for val in relation.relationships.__dict__[relationship_type]]
+
+            # determine downstream relations from relation patterns
+            downstream_relations = set(
+                filter(
+                    lambda x: single_full_pattern_match(x, relation_pattern_dict),  # noqa  pylint: disable=cell-var-from-loop
+                    available_nodes
+                )
+            )
+            if not downstream_relations:
+                raise InvalidRelationshipException(
+                    f'Relationship {relation_pattern_dict} was specified, '
+                    f'but does not match any relations. '
+                    f'Please verify replica configuration.'
+                )
+
+            # for each relationship, find up/downstream relations, then create the appropriate edges
+            for relationship in relationship_dicts:
+                # check for wild cards
+                possible_wildcard_attrs = ('database', 'schema',)
+                wildcard_attrs = [attr for attr in possible_wildcard_attrs if relationship[attr] is None]
+
+                # if there are wildcard attributes, partition downstream relations
+                if wildcard_attrs:
+                    wildcard_partitions = {}
+                    for down_rel in downstream_relations:
+                        # create wildcard key off of (in case there are multiple wildcards)
+                        wildcard_key = '|'.join([getattr(down_rel, attr) for attr in wildcard_attrs])
+                        val = wildcard_partitions.get(wildcard_key, [])
+                        val.append(down_rel)
+                        wildcard_partitions[wildcard_key] = val
+
+                    for _, downstream_partition in wildcard_partitions.items():
+                        # populate any wildcard patterns with the appropriate values from first element
+                        for attr in wildcard_attrs:
+                            relationship[attr] = getattr(downstream_partition[0], attr)
+
+                        graph = SnowShuGraph._process_downstream_relation_set(relationship,
+                                                                              downstream_relations,
+                                                                              graph,
+                                                                              available_nodes)
+                # no wildcards present in relationship definition
+                else:
```

Comment:
could you remove the `else` and only have one call to `SnowShuGraph._process_downstream_relation_set()`

Reply:
No, but that's because I screwed up line 170. The other call needs to use `downstream_partition` not `downstream_relations` to handle the wildcard behavior properly

Reply:
Fixed with test.